include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -0,0 +1,454 @@
+};
+
+
+/**
+ * This mixin provides apply and common iterative solver functionality to all
+ * the batched solvers.
+ *
+ * @tparam ConcreteSolver  The concrete solver class.
+ * @tparam ValueType  The value type of the multivectors.
+ * @tparam PolymorphicBase  The base class; must be a subclass of BatchLinOp.
+ */
+template <typename ConcreteSolver,
+          typename ValueType = typename ConcreteSolver::value_type,
+          typename PolymorphicBase = BatchLinOp>
+class EnableBatchSolver
+    : public BatchSolver,
+      public EnableBatchLinOp<ConcreteSolver, PolymorphicBase> {
+public:
+    using real_type = remove_complex<ValueType>;
+    const ConcreteSolver* apply(ptr_param<const MultiVector<ValueType>> b,
+                                ptr_param<MultiVector<ValueType>> x) const
+    {
+        this->validate_application_parameters(b.get(), x.get());
+        auto exec = this->get_executor();
+        this->apply_impl(make_temporary_clone(exec, b).get(),
+                         make_temporary_clone(exec, x).get());
+        return self();
+    }
+
+    const ConcreteSolver* apply(ptr_param<const MultiVector<ValueType>> alpha,
+                                ptr_param<const MultiVector<ValueType>> b,
+                                ptr_param<const MultiVector<ValueType>> beta,
+                                ptr_param<MultiVector<ValueType>> x) const
+    {
+        this->validate_application_parameters(alpha.get(), b.get(), beta.get(),
+                                              x.get());
+        auto exec = this->get_executor();
+        this->apply_impl(make_temporary_clone(exec, alpha).get(),
+                         make_temporary_clone(exec, b).get(),
+                         make_temporary_clone(exec, beta).get(),
+                         make_temporary_clone(exec, x).get());
+        return self();
+    }
+
+    ConcreteSolver* apply(ptr_param<const MultiVector<ValueType>> b,
+                          ptr_param<MultiVector<ValueType>> x)
+    {
+        static_cast<const ConcreteSolver*>(this)->apply(b, x);
+        return self();
+    }
+
+    ConcreteSolver* apply(ptr_param<const MultiVector<ValueType>> alpha,
+                          ptr_param<const MultiVector<ValueType>> b,
+                          ptr_param<const MultiVector<ValueType>> beta,
+                          ptr_param<MultiVector<ValueType>> x)
+    {
+        static_cast<const ConcreteSolver*>(this)->apply(alpha, b, beta, x);
+        return self();
+    }
+
+protected:
+    GKO_ENABLE_SELF(ConcreteSolver);
+
+    explicit EnableBatchSolver(std::shared_ptr<const Executor> exec)
+        : EnableBatchLinOp<ConcreteSolver, PolymorphicBase>(std::move(exec))
+    {}
+
+    template <typename FactoryParameters>
+    explicit EnableBatchSolver(std::shared_ptr<const Executor> exec,
+                               std::shared_ptr<const BatchLinOp> system_matrix,
+                               const FactoryParameters& params)
```

Comment:
```suggestion
    explicit EnableBatchSolver(std::shared_ptr<const Executor> exec,
                               std::shared_ptr<const BatchLinOp> system_matrix,
                               const typename ConcreteSolver::parameters_type & params)
```

Reply:
Any reason you prefer this version ?

Reply:
The original interface seems to accept any kind of parameters not from concrete solver?

Reply:
I think it makes it easier if the type of parameters (`parameters_type` now) changes in the future.  Implicitly the `params` parameter needs to have the necessary members anyway and that is checked at compile time.